jolt-core/src/jolt/vm/ram.rs

```diff
@@ -411,181 +457,236 @@ impl<F: JoltField, ProofTranscript: Transcript> BatchableSumcheckInstance<F, Pro
-                    B_evals[2] * inner_sum[2],
-                ]
+                // Multiply by B evaluations
+                for j in 0..degree {
+                    evals[j] *= B_evals[j];
+                }
+                evals
             })
             .reduce(
-                || [F::zero(); 3],
-                |running, new| {
-                    [
-                        running[0] + new[0],
-                        running[1] + new[1],
-                        running[2] + new[2],
-                    ]
+                || vec![F::zero(); degree],
+                |mut running, new| {
+                    for j in 0..degree {
+                        running[j] += new[j];
+                    }
+                    running
                 },
-            );
+            )
+            .into_iter()
+            .enumerate()
+            .for_each(|(i, val)| univariate_poly_evals[i] = val);
 
-        univariate_poly_evals.to_vec()
+        univariate_poly_evals
     }
 
     /// Compute prover message for phase 2 (last log(T) rounds)
     fn compute_phase2_message(&self, _round: usize) -> Vec<F> {
-        const DEGREE: usize = 3;
         let prover_state = self
             .prover_state
             .as_ref()
             .expect("Prover state not initialized");
-        let h_poly = prover_state.H.as_ref().expect("H not initialized");
+        let h_polys = prover_state
+            .H
+            .as_ref()
+            .expect("H polynomials not initialized");
+        let d = prover_state.d;
+        let degree = 2 + d;
 
-        let mut univariate_poly_evals: [F; 3] = (0..prover_state.D.len() / 2)
+        let mut univariate_poly_evals = vec![F::zero(); degree];
+
+        (0..prover_state.D.len() / 2)
             .into_par_iter()
             .map(|i| {
                 let D_evals = prover_state
                     .D
-                    .sumcheck_evals(i, DEGREE, BindingOrder::LowToHigh);
-                let H_evals = h_poly.sumcheck_evals(i, DEGREE, BindingOrder::LowToHigh);
+                    .sumcheck_evals(i, degree, BindingOrder::LowToHigh);
 
-                [
-                    D_evals[0] * (H_evals[0] * H_evals[0] - H_evals[0]),
-                    D_evals[1] * (H_evals[1] * H_evals[1] - H_evals[1]),
-                    D_evals[2] * (H_evals[2] * H_evals[2] - H_evals[2]),
-                ]
+                let mut evals = vec![F::zero(); degree];
+
+                // For each polynomial in the batch
+                for j in 0..d {
+                    let H_j_evals = h_polys[j].sumcheck_evals(i, degree, BindingOrder::LowToHigh);
+
+                    // For each evaluation point
+                    for k in 0..degree {
+                        // Add z^j * (H_j^2 - H_j) * D
+                        evals[k] += prover_state.z_powers[j]
+                            * D_evals[k]
+                            * (H_j_evals[k] * H_j_evals[k] - H_j_evals[k]);
```

Comment:
while you're refactoring, might as well change this
```suggestion
                            * (H_j_evals[k].square() - H_j_evals[k]);
```